File: ClientsToAnalyze/Android/1.5.0.45-165.VoiceChat/app/src/main/java/com/google/protobuf/UnknownFieldSet.java
package com.google.protobuf;

import com.google.protobuf.AbstractMessageLite;
import com.google.protobuf.ByteString;
import com.google.protobuf.MessageLite;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/* loaded from: classes.dex */
public final class UnknownFieldSet implements MessageLite {
    private static final UnknownFieldSet defaultInstance = new UnknownFieldSet(Collections.emptyMap());
    private Map<Integer, Field> fields;

    private UnknownFieldSet() {
    }

    public static Builder newBuilder() {
        return Builder.create();
    }

    public static Builder newBuilder(UnknownFieldSet copyFrom) {
        return newBuilder().mergeFrom(copyFrom);
    }

    public static UnknownFieldSet getDefaultInstance() {
        return defaultInstance;
    }

    @Override // com.google.protobuf.MessageLiteOrBuilder, com.google.protobuf.MessageOrBuilder
    public UnknownFieldSet getDefaultInstanceForType() {
        return defaultInstance;
    }

    private UnknownFieldSet(Map<Integer, Field> fields) {
        this.fields = fields;
    }

    /* synthetic */ UnknownFieldSet(Map map, UnknownFieldSet unknownFieldSet) {
        this(map);
    }

    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return (other instanceof UnknownFieldSet) && this.fields.equals(((UnknownFieldSet) other).fields);
    }

    public int hashCode() {
        return this.fields.hashCode();
    }

    public Map<Integer, Field> asMap() {
        return this.fields;
    }

    public boolean hasField(int number) {
        return this.fields.containsKey(Integer.valueOf(number));
    }

    public Field getField(int number) {
        Field result = this.fields.get(Integer.valueOf(number));
        return result == null ? Field.getDefaultInstance() : result;
    }

    @Override // com.google.protobuf.MessageLite
    public void writeTo(CodedOutputStream output) throws IOException {
        for (Map.Entry<Integer, Field> entry : this.fields.entrySet()) {
            entry.getValue().writeTo(entry.getKey().intValue(), output);
        }
    }

    public String toString() {
        return TextFormat.printToString(this);
    }

    @Override // com.google.protobuf.MessageLite
    public ByteString toByteString() {
        try {
            ByteString.CodedBuilder out = ByteString.newCodedBuilder(getSerializedSize());
            writeTo(out.getCodedOutput());
            return out.build();
        } catch (IOException e) {
            throw new RuntimeException("Serializing to a ByteString threw an IOException (should never happen).", e);
        }
    }

    @Override // com.google.protobuf.MessageLite
    public byte[] toByteArray() {
        try {
            byte[] result = new byte[getSerializedSize()];
            CodedOutputStream output = CodedOutputStream.newInstance(result);
            writeTo(output);
            output.checkNoSpaceLeft();
            return result;
        } catch (IOException e) {
            throw new RuntimeException("Serializing to a byte array threw an IOException (should never happen).", e);
        }
    }

    @Override // com.google.protobuf.MessageLite
    public void writeTo(OutputStream output) throws IOException {
        CodedOutputStream codedOutput = CodedOutputStream.newInstance(output);
        writeTo(codedOutput);
        codedOutput.flush();
    }

    @Override // com.google.protobuf.MessageLite
    public void writeDelimitedTo(OutputStream output) throws IOException {
        CodedOutputStream codedOutput = CodedOutputStream.newInstance(output);
        codedOutput.writeRawVarint32(getSerializedSize());
        writeTo(codedOutput);
        codedOutput.flush();
    }

    @Override // com.google.protobuf.MessageLite
    public int getSerializedSize() {
        int result = 0;
        for (Map.Entry<Integer, Field> entry : this.fields.entrySet()) {
            result += entry.getValue().getSerializedSize(entry.getKey().intValue());
        }
        return result;
    }

    public void writeAsMessageSetTo(CodedOutputStream output) throws IOException {
        for (Map.Entry<Integer, Field> entry : this.fields.entrySet()) {
            entry.getValue().writeAsMessageSetExtensionTo(entry.getKey().intValue(), output);
        }
    }

    public int getSerializedSizeAsMessageSet() {
        int result = 0;
        for (Map.Entry<Integer, Field> entry : this.fields.entrySet()) {
            result += entry.getValue().getSerializedSizeAsMessageSetExtension(entry.getKey().intValue());
        }
        return result;
    }

    @Override // com.google.protobuf.MessageLiteOrBuilder
    public boolean isInitialized() {
        return true;
    }

    public static UnknownFieldSet parseFrom(CodedInputStream input) throws IOException {
        return newBuilder().mergeFrom(input).build();
    }

    public static UnknownFieldSet parseFrom(ByteString data) throws InvalidProtocolBufferException {
        return newBuilder().mergeFrom(data).build();
    }

    public static UnknownFieldSet parseFrom(byte[] data) throws InvalidProtocolBufferException {
        return newBuilder().mergeFrom(data).build();
    }

    public static UnknownFieldSet parseFrom(InputStream input) throws IOException {
        return newBuilder().mergeFrom(input).build();
    }

    @Override // com.google.protobuf.MessageLite, com.google.protobuf.Message
    public Builder newBuilderForType() {
        return newBuilder();
    }

    @Override // com.google.protobuf.MessageLite, com.google.protobuf.Message
    public Builder toBuilder() {
        return newBuilder().mergeFrom(this);
    }

    public static final class Builder implements MessageLite.Builder {
        private Map<Integer, Field> fields;
        private Field.Builder lastField;
        private int lastFieldNumber;

        private Builder() {
        }

        /* JADX INFO: Access modifiers changed from: private */
        public static Builder create() {
            Builder builder = new Builder();
            builder.reinitialize();
            return builder;
        }

        private Field.Builder getFieldBuilder(int number) {
            if (this.lastField != null) {
                if (number == this.lastFieldNumber) {
                    return this.lastField;
                }
                addField(this.lastFieldNumber, this.lastField.build());
            }
            if (number == 0) {
                return null;
            }
            Field existing = this.fields.get(Integer.valueOf(number));
            this.lastFieldNumber = number;
            this.lastField = Field.newBuilder();
            if (existing != null) {
                this.lastField.mergeFrom(existing);
            }
            return this.lastField;
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public UnknownFieldSet build() {
            UnknownFieldSet result;
            UnknownFieldSet unknownFieldSet = null;
            getFieldBuilder(0);
            if (this.fields.isEmpty()) {
                result = UnknownFieldSet.getDefaultInstance();
            } else {
                result = new UnknownFieldSet(Collections.unmodifiableMap(this.fields), unknownFieldSet);
            }
            this.fields = null;
            return result;
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public UnknownFieldSet buildPartial() {
            return build();
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        /* renamed from: clone, reason: merged with bridge method [inline-methods] */
        public Builder m3102clone() {
            getFieldBuilder(0);
            return UnknownFieldSet.newBuilder().mergeFrom(new UnknownFieldSet(this.fields, null));
        }

        @Override // com.google.protobuf.MessageLiteOrBuilder, com.google.protobuf.MessageOrBuilder
        public UnknownFieldSet getDefaultInstanceForType() {
            return UnknownFieldSet.getDefaultInstance();
        }

        private void reinitialize() {
            this.fields = Collections.emptyMap();
            this.lastFieldNumber = 0;
            this.lastField = null;
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder clear() {
            reinitialize();
            return this;
        }

        public Builder mergeFrom(UnknownFieldSet other) {
            if (other != UnknownFieldSet.getDefaultInstance()) {
                for (Map.Entry<Integer, Field> entry : other.fields.entrySet()) {
                    mergeField(entry.getKey().intValue(), entry.getValue());
                }
            }
            return this;
        }

        public Builder mergeField(int number, Field field) {
            if (number == 0) {
                throw new IllegalArgumentException("Zero is not a valid field number.");
            }
            if (hasField(number)) {
                getFieldBuilder(number).mergeFrom(field);
            } else {
                addField(number, field);
            }
            return this;
        }

        public Builder mergeVarintField(int number, int value) {
            if (number == 0) {
                throw new IllegalArgumentException("Zero is not a valid field number.");
            }
            getFieldBuilder(number).addVarint(value);
            return this;
        }

        public boolean hasField(int number) {
            if (number == 0) {
                throw new IllegalArgumentException("Zero is not a valid field number.");
            }
            return number == this.lastFieldNumber || this.fields.containsKey(Integer.valueOf(number));
        }

        public Builder addField(int number, Field field) {
            if (number == 0) {
                throw new IllegalArgumentException("Zero is not a valid field number.");
            }
            if (this.lastField != null && this.lastFieldNumber == number) {
                this.lastField = null;
                this.lastFieldNumber = 0;
            }
            if (this.fields.isEmpty()) {
                this.fields = new TreeMap();
            }
            this.fields.put(Integer.valueOf(number), field);
            return this;
        }

        public Map<Integer, Field> asMap() {
            getFieldBuilder(0);
            return Collections.unmodifiableMap(this.fields);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(CodedInputStream input) throws IOException {
            int tag;
            do {
                tag = input.readTag();
                if (tag == 0) {
                    break;
                }
            } while (mergeFieldFrom(tag, input));
            return this;
        }

        public boolean mergeFieldFrom(int tag, CodedInputStream input) throws IOException {
            int number = WireFormat.getTagFieldNumber(tag);
            switch (WireFormat.getTagWireType(tag)) {
                case 0:
                    getFieldBuilder(number).addVarint(input.readInt64());
                    return true;
                case 1:
                    getFieldBuilder(number).addFixed64(input.readFixed64());
                    return true;
                case 2:
                    getFieldBuilder(number).addLengthDelimited(input.readBytes());
                    return true;
                case 3:
                    Builder subBuilder = UnknownFieldSet.newBuilder();
                    input.readGroup(number, subBuilder, ExtensionRegistry.getEmptyRegistry());
                    getFieldBuilder(number).addGroup(subBuilder.build());
                    return true;
                case 4:
                    return false;
                case 5:
                    getFieldBuilder(number).addFixed32(input.readFixed32());
                    return true;
                default:
                    throw InvalidProtocolBufferException.invalidWireType();
            }
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(ByteString data) throws InvalidProtocolBufferException {
            try {
                CodedInputStream input = data.newCodedInput();
                mergeFrom(input);
                input.checkLastTagWas(0);
                return this;
            } catch (InvalidProtocolBufferException e) {
                throw e;
            } catch (IOException e2) {
                throw new RuntimeException("Reading from a ByteString threw an IOException (should never happen).", e2);
            }
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(byte[] data) throws InvalidProtocolBufferException {
            try {
                CodedInputStream input = CodedInputStream.newInstance(data);
                mergeFrom(input);
                input.checkLastTagWas(0);
                return this;
            } catch (InvalidProtocolBufferException e) {
                throw e;
            } catch (IOException e2) {
                throw new RuntimeException("Reading from a byte array threw an IOException (should never happen).", e2);
            }
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(InputStream input) throws IOException {
            CodedInputStream codedInput = CodedInputStream.newInstance(input);
            mergeFrom(codedInput);
            codedInput.checkLastTagWas(0);
            return this;
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public boolean mergeDelimitedFrom(InputStream input) throws IOException {
            int firstByte = input.read();
            if (firstByte == -1) {
                return false;
            }
            int size = CodedInputStream.readRawVarint32(firstByte, input);
            InputStream limitedInput = new AbstractMessageLite.Builder.LimitedInputStream(input, size);
            mergeFrom(limitedInput);
            return true;
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public boolean mergeDelimitedFrom(InputStream input, ExtensionRegistryLite extensionRegistry) throws IOException {
            return mergeDelimitedFrom(input);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(CodedInputStream input, ExtensionRegistryLite extensionRegistry) throws IOException {
            return mergeFrom(input);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(ByteString data, ExtensionRegistryLite extensionRegistry) throws InvalidProtocolBufferException {
            return mergeFrom(data);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(byte[] data, int off, int len) throws InvalidProtocolBufferException {
            try {
                CodedInputStream input = CodedInputStream.newInstance(data, off, len);
                mergeFrom(input);
                input.checkLastTagWas(0);
                return this;
            } catch (InvalidProtocolBufferException e) {
                throw e;
            } catch (IOException e2) {
                throw new RuntimeException("Reading from a byte array threw an IOException (should never happen).", e2);
            }
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(byte[] data, ExtensionRegistryLite extensionRegistry) throws InvalidProtocolBufferException {
            return mergeFrom(data);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(byte[] data, int off, int len, ExtensionRegistryLite extensionRegistry) throws InvalidProtocolBufferException {
            return mergeFrom(data, off, len);
        }

        @Override // com.google.protobuf.MessageLite.Builder, com.google.protobuf.Message.Builder
        public Builder mergeFrom(InputStream input, ExtensionRegistryLite extensionRegistry) throws IOException {
            return mergeFrom(input);
        }

        @Override // com.google.protobuf.MessageLiteOrBuilder
        public boolean isInitialized() {
            return true;
        }
    }

    public static final class Field {
        private static final Field fieldDefaultInstance = newBuilder().build();
        private List<Integer> fixed32;
        private List<Long> fixed64;
        private List<UnknownFieldSet> group;
        private List<ByteString> lengthDelimited;
        private List<Long> varint;

        private Field() {
        }

        /* synthetic */ Field(Field field) {
            this();
        }

        public static Builder newBuilder() {
            return Builder.create();
        }

        public static Builder newBuilder(Field copyFrom) {
            return newBuilder().mergeFrom(copyFrom);
        }

        public static Field getDefaultInstance() {
            return fieldDefaultInstance;
        }

        public List<Long> getVarintList() {
            return this.varint;
        }

        public List<Integer> getFixed32List() {
            return this.fixed32;
        }

        public List<Long> getFixed64List() {
            return this.fixed64;
        }

        public List<ByteString> getLengthDelimitedList() {
            return this.lengthDelimited;
        }

        public List<UnknownFieldSet> getGroupList() {
            return this.group;
        }

        public boolean equals(Object other) {
            if (this == other) {
                return true;
            }
            if (!(other instanceof Field)) {
                return false;
            }
            return Arrays.equals(getIdentityArray(), ((Field) other).getIdentityArray());
        }

        public int hashCode() {
            return Arrays.hashCode(getIdentityArray());
        }

        private Object[] getIdentityArray() {
            return new Object[]{this.varint, this.fixed32, this.fixed64, this.lengthDelimited, this.group};
        }

        public void writeTo(int fieldNumber, CodedOutputStream output) throws IOException {
            Iterator<Long> it = this.varint.iterator();
            while (it.hasNext()) {
                long value = it.next().longValue();
                output.writeUInt64(fieldNumber, value);
            }
            Iterator<Integer> it2 = this.fixed32.iterator();
            while (it2.hasNext()) {
                int value2 = it2.next().intValue();
                output.writeFixed32(fieldNumber, value2);
            }
            Iterator<Long> it3 = this.fixed64.iterator();
            while (it3.hasNext()) {
                long value3 = it3.next().longValue();
                output.writeFixed64(fieldNumber, value3);
            }
            for (ByteString value4 : this.lengthDelimited) {
                output.writeBytes(fieldNumber, value4);
            }
            for (UnknownFieldSet value5 : this.group) {
                output.writeGroup(fieldNumber, value5);
            }
        }

        public int getSerializedSize(int fieldNumber) {
            int result = 0;
            Iterator<Long> it = this.varint.iterator();
            while (it.hasNext()) {
                long value = it.next().longValue();
                result += CodedOutputStream.computeUInt64Size(fieldNumber, value);
            }
            Iterator<Integer> it2 = this.fixed32.iterator();
            while (it2.hasNext()) {
                int value2 = it2.next().intValue();
                result += CodedOutputStream.computeFixed32Size(fieldNumber, value2);
            }
            Iterator<Long> it3 = this.fixed64.iterator();
            while (it3.hasNext()) {
                long value3 = it3.next().longValue();
                result += CodedOutputStream.computeFixed64Size(fieldNumber, value3);
            }
            for (ByteString value4 : this.lengthDelimited) {
                result += CodedOutputStream.computeBytesSize(fieldNumber, value4);
            }
            for (UnknownFieldSet value5 : this.group) {
                result += CodedOutputStream.computeGroupSize(fieldNumber, value5);
            }
            return result;
        }

        public void writeAsMessageSetExtensionTo(int fieldNumber, CodedOutputStream output) throws IOException {
            for (ByteString value : this.lengthDelimited) {
                output.writeRawMessageSetExtension(fieldNumber, value);
            }
        }

        public int getSerializedSizeAsMessageSetExtension(int fieldNumber) {
            int result = 0;
            for (ByteString value : this.lengthDelimited) {
                result += CodedOutputStream.computeRawMessageSetExtensionSize(fieldNumber, value);
            }
            return result;
        }

        public static final class Builder {
            private Field result;

            private Builder() {
            }

            /* JADX INFO: Access modifiers changed from: private */
            public static Builder create() {
                Builder builder = new Builder();
                builder.result = new Field(null);
                return builder;
            }

            public Field build() {
                if (this.result.varint == null) {
                    this.result.varint = Collections.emptyList();
                } else {
                    this.result.varint = Collections.unmodifiableList(this.result.varint);
                }
                if (this.result.fixed32 == null) {
                    this.result.fixed32 = Collections.emptyList();
                } else {
                    this.result.fixed32 = Collections.unmodifiableList(this.result.fixed32);
                }
                if (this.result.fixed64 == null) {
                    this.result.fixed64 = Collections.emptyList();
                } else {
                    this.result.fixed64 = Collections.unmodifiableList(this.result.fixed64);
                }
                if (this.result.lengthDelimited == null) {
                    this.result.lengthDelimited = Collections.emptyList();
                } else {
                    this.result.lengthDelimited = Collections.unmodifiableList(this.result.lengthDelimited);
                }
                if (this.result.group == null) {
                    this.result.group = Collections.emptyList();
                } else {
                    this.result.group = Collections.unmodifiableList(this.result.group);
                }
                Field returnMe = this.result;
                this.result = null;
                return returnMe;
            }

            public Builder clear() {
                this.result = new Field(null);
                return this;
            }

            public Builder mergeFrom(Field other) {
                if (!other.varint.isEmpty()) {
                    if (this.result.varint == null) {
                        this.result.varint = new ArrayList();
                    }
                    this.result.varint.addAll(other.varint);
                }
                if (!other.fixed32.isEmpty()) {
                    if (this.result.fixed32 == null) {
                        this.result.fixed32 = new ArrayList();
                    }
                    this.result.fixed32.addAll(other.fixed32);
                }
                if (!other.fixed64.isEmpty()) {
                    if (this.result.fixed64 == null) {
                        this.result.fixed64 = new ArrayList();
                    }
                    this.result.fixed64.addAll(other.fixed64);
                }
                if (!other.lengthDelimited.isEmpty()) {
                    if (this.result.lengthDelimited == null) {
                        this.result.lengthDelimited = new ArrayList();
                    }
                    this.result.lengthDelimited.addAll(other.lengthDelimited);
                }
                if (!other.group.isEmpty()) {
                    if (this.result.group == null) {
                        this.result.group = new ArrayList();
                    }
                    this.result.group.addAll(other.group);
                }
                return this;
            }

            public Builder addVarint(long value) {
                if (this.result.varint == null) {
                    this.result.varint = new ArrayList();
                }
                this.result.varint.add(Long.valueOf(value));
                return this;
            }

            public Builder addFixed32(int value) {
                if (this.result.fixed32 == null) {
                    this.result.fixed32 = new ArrayList();
                }
                this.result.fixed32.add(Integer.valueOf(value));
                return this;
            }

            public Builder addFixed64(long value) {
                if (this.result.fixed64 == null) {
                    this.result.fixed64 = new ArrayList();
                }
                this.result.fixed64.add(Long.valueOf(value));
                return this;
            }

            public Builder addLengthDelimited(ByteString value) {
                if (this.result.lengthDelimited == null) {
                    this.result.lengthDelimited = new ArrayList();
                }
                this.result.lengthDelimited.add(value);
                return this;
            }

            public Builder addGroup(UnknownFieldSet value) {
                if (this.result.group == null) {
                    this.result.group = new ArrayList();
                }
                this.result.group.add(value);
                return this;
            }
        }
    }
}

File: ClientsToAnalyze/WindowsPhone78/Decomp/Google.ProtocolBuffers/Google.ProtocolBuffers/UnknownFieldSet.cs
using System;
using System.Collections.Generic;
using System.IO;
using Google.ProtocolBuffers.Collections;
using Google.ProtocolBuffers.Descriptors;

namespace Google.ProtocolBuffers;

public sealed class UnknownFieldSet
{
	public sealed class Builder
	{
		private IDictionary<int, UnknownField> fields = new SortedList<int, UnknownField>();

		private int lastFieldNumber;

		private UnknownField.Builder lastField;

		internal Builder()
		{
		}

		private UnknownField.Builder GetFieldBuilder(int number)
		{
			if (lastField != null)
			{
				if (number == lastFieldNumber)
				{
					return lastField;
				}
				AddField(lastFieldNumber, lastField.Build());
			}
			if (number == 0)
			{
				return null;
			}
			lastField = UnknownField.CreateBuilder();
			if (fields.TryGetValue(number, out var value))
			{
				lastField.MergeFrom(value);
			}
			lastFieldNumber = number;
			return lastField;
		}

		public UnknownFieldSet Build()
		{
			GetFieldBuilder(0);
			UnknownFieldSet result = ((fields.Count == 0) ? DefaultInstance : new UnknownFieldSet(fields));
			fields = null;
			return result;
		}

		public Builder AddField(int number, UnknownField field)
		{
			if (number == 0)
			{
				throw new ArgumentOutOfRangeException("number", "Zero is not a valid field number.");
			}
			if (lastField != null && lastFieldNumber == number)
			{
				lastField = null;
				lastFieldNumber = 0;
			}
			fields[number] = field;
			return this;
		}

		public Builder Clear()
		{
			fields.Clear();
			lastFieldNumber = 0;
			lastField = null;
			return this;
		}

		public Builder MergeFrom(CodedInputStream input)
		{
			uint num;
			do
			{
				num = input.ReadTag();
			}
			while (num != 0 && MergeFieldFrom(num, input));
			return this;
		}

		[CLSCompliant(false)]
		public bool MergeFieldFrom(uint tag, CodedInputStream input)
		{
			int tagFieldNumber = WireFormat.GetTagFieldNumber(tag);
			switch (WireFormat.GetTagWireType(tag))
			{
			case WireFormat.WireType.Varint:
				GetFieldBuilder(tagFieldNumber).AddVarint(input.ReadUInt64());
				return true;
			case WireFormat.WireType.Fixed64:
				GetFieldBuilder(tagFieldNumber).AddFixed64(input.ReadFixed64());
				return true;
			case WireFormat.WireType.LengthDelimited:
				GetFieldBuilder(tagFieldNumber).AddLengthDelimited(input.ReadBytes());
				return true;
			case WireFormat.WireType.StartGroup:
			{
				Builder builder = CreateBuilder();
				input.ReadUnknownGroup(tagFieldNumber, builder);
				GetFieldBuilder(tagFieldNumber).AddGroup(builder.Build());
				return true;
			}
			case WireFormat.WireType.EndGroup:
				return false;
			case WireFormat.WireType.Fixed32:
				GetFieldBuilder(tagFieldNumber).AddFixed32(input.ReadFixed32());
				return true;
			default:
				throw InvalidProtocolBufferException.InvalidWireType();
			}
		}

		public Builder MergeFrom(Stream input)
		{
			CodedInputStream codedInputStream = CodedInputStream.CreateInstance(input);
			MergeFrom(codedInputStream);
			codedInputStream.CheckLastTagWas(0u);
			return this;
		}

		public Builder MergeFrom(ByteString data)
		{
			CodedInputStream codedInputStream = data.CreateCodedInput();
			MergeFrom(codedInputStream);
			codedInputStream.CheckLastTagWas(0u);
			return this;
		}

		public Builder MergeFrom(byte[] data)
		{
			CodedInputStream codedInputStream = CodedInputStream.CreateInstance(data);
			MergeFrom(codedInputStream);
			codedInputStream.CheckLastTagWas(0u);
			return this;
		}

		[CLSCompliant(false)]
		public Builder MergeVarintField(int number, ulong value)
		{
			if (number == 0)
			{
				throw new ArgumentOutOfRangeException("number", "Zero is not a valid field number.");
			}
			GetFieldBuilder(number).AddVarint(value);
			return this;
		}

		public Builder MergeFrom(UnknownFieldSet other)
		{
			if (other != DefaultInstance)
			{
				foreach (KeyValuePair<int, UnknownField> field in other.fields)
				{
					MergeField(field.Key, field.Value);
				}
			}
			return this;
		}

		public bool HasField(int number)
		{
			if (number == 0)
			{
				throw new ArgumentOutOfRangeException("number", "Zero is not a valid field number.");
			}
			if (number != lastFieldNumber)
			{
				return fields.ContainsKey(number);
			}
			return true;
		}

		public Builder MergeField(int number, UnknownField field)
		{
			if (number == 0)
			{
				throw new ArgumentOutOfRangeException("number", "Zero is not a valid field number.");
			}
			if (HasField(number))
			{
				GetFieldBuilder(number).MergeFrom(field);
			}
			else
			{
				AddField(number, field);
			}
			return this;
		}

		internal void MergeFrom(CodedInputStream input, ExtensionRegistry extensionRegistry, IBuilder builder)
		{
			uint num;
			do
			{
				num = input.ReadTag();
			}
			while (num != 0 && MergeFieldFrom(input, extensionRegistry, builder, num));
		}

		internal bool MergeFieldFrom(CodedInputStream input, ExtensionRegistry extensionRegistry, IBuilder builder, uint tag)
		{
			MessageDescriptor descriptorForType = builder.DescriptorForType;
			if (descriptorForType.Options.MessageSetWireFormat && tag == WireFormat.MessageSetTag.ItemStart)
			{
				MergeMessageSetExtensionFromCodedStream(input, extensionRegistry, builder);
				return true;
			}
			WireFormat.WireType tagWireType = WireFormat.GetTagWireType(tag);
			int tagFieldNumber = WireFormat.GetTagFieldNumber(tag);
			IMessage message = null;
			FieldDescriptor fieldDescriptor;
			if (descriptorForType.IsExtensionNumber(tagFieldNumber))
			{
				ExtensionInfo extensionInfo = extensionRegistry[descriptorForType, tagFieldNumber];
				if (extensionInfo == null)
				{
					fieldDescriptor = null;
				}
				else
				{
					fieldDescriptor = extensionInfo.Descriptor;
					message = extensionInfo.DefaultInstance;
				}
			}
			else
			{
				fieldDescriptor = descriptorForType.FindFieldByNumber(tagFieldNumber);
			}
			if (fieldDescriptor == null || tagWireType != WireFormat.GetWireType(fieldDescriptor))
			{
				return MergeFieldFrom(tag, input);
			}
			if (fieldDescriptor.IsPacked)
			{
				int byteLimit = (int)input.ReadRawVarint32();
				int oldLimit = input.PushLimit(byteLimit);
				if (fieldDescriptor.FieldType == FieldType.Enum)
				{
					while (!input.ReachedLimit)
					{
						int number = input.ReadEnum();
						object obj = fieldDescriptor.EnumType.FindValueByNumber(number);
						if (obj == null)
						{
							return true;
						}
						builder.WeakAddRepeatedField(fieldDescriptor, obj);
					}
				}
				else
				{
					while (!input.ReachedLimit)
					{
						object value = input.ReadPrimitiveField(fieldDescriptor.FieldType);
						builder.WeakAddRepeatedField(fieldDescriptor, value);
					}
				}
				input.PopLimit(oldLimit);
			}
			else
			{
				object obj2;
				switch (fieldDescriptor.FieldType)
				{
				case FieldType.Group:
				case FieldType.Message:
				{
					IBuilder builder2 = ((message == null) ? builder.CreateBuilderForField(fieldDescriptor) : message.WeakCreateBuilderForType());
					if (!fieldDescriptor.IsRepeated)
					{
						builder2.WeakMergeFrom((IMessage)builder[fieldDescriptor]);
					}
					if (fieldDescriptor.FieldType == FieldType.Group)
					{
						input.ReadGroup(fieldDescriptor.FieldNumber, builder2, extensionRegistry);
					}
					else
					{
						input.ReadMessage(builder2, extensionRegistry);
					}
					obj2 = builder2.WeakBuild();
					break;
				}
				case FieldType.Enum:
				{
					int num = input.ReadEnum();
					obj2 = fieldDescriptor.EnumType.FindValueByNumber(num);
					if (obj2 == null)
					{
						MergeVarintField(tagFieldNumber, (ulong)num);
						return true;
					}
					break;
				}
				default:
					obj2 = input.ReadPrimitiveField(fieldDescriptor.FieldType);
					break;
				}
				if (fieldDescriptor.IsRepeated)
				{
					builder.WeakAddRepeatedField(fieldDescriptor, obj2);
				}
				else
				{
					builder[fieldDescriptor] = obj2;
				}
			}
			return true;
		}

		private void MergeMessageSetExtensionFromCodedStream(CodedInputStream input, ExtensionRegistry extensionRegistry, IBuilder builder)
		{
			MessageDescriptor descriptorForType = builder.DescriptorForType;
			int num = 0;
			ByteString byteString = null;
			IBuilder builder2 = null;
			FieldDescriptor field = null;
			while (true)
			{
				uint num2 = input.ReadTag();
				if (num2 == 0)
				{
					break;
				}
				if (num2 == WireFormat.MessageSetTag.TypeID)
				{
					num = input.ReadInt32();
					if (num == 0)
					{
						continue;
					}
					ExtensionInfo extensionInfo = extensionRegistry[descriptorForType, num];
					if (extensionInfo != null)
					{
						field = extensionInfo.Descriptor;
						builder2 = extensionInfo.DefaultInstance.WeakCreateBuilderForType();
						IMessage message = (IMessage)builder[field];
						if (message != null)
						{
							builder2.WeakMergeFrom(message);
						}
						if (byteString != null)
						{
							builder2.WeakMergeFrom(byteString.CreateCodedInput());
							byteString = null;
						}
					}
					else if (byteString != null)
					{
						MergeField(num, UnknownField.CreateBuilder().AddLengthDelimited(byteString).Build());
						byteString = null;
					}
				}
				else if (num2 == WireFormat.MessageSetTag.Message)
				{
					if (num == 0)
					{
						byteString = input.ReadBytes();
					}
					else if (builder2 == null)
					{
						MergeField(num, UnknownField.CreateBuilder().AddLengthDelimited(input.ReadBytes()).Build());
					}
					else
					{
						input.ReadMessage(builder2, extensionRegistry);
					}
				}
				else if (!input.SkipField(num2))
				{
					break;
				}
			}
			input.CheckLastTagWas(WireFormat.MessageSetTag.ItemEnd);
			if (builder2 != null)
			{
				builder[field] = builder2.WeakBuild();
			}
		}
	}

	private static readonly UnknownFieldSet defaultInstance = new UnknownFieldSet(new Dictionary<int, UnknownField>());

	private readonly IDictionary<int, UnknownField> fields;

	public static UnknownFieldSet DefaultInstance => defaultInstance;

	public IDictionary<int, UnknownField> FieldDictionary => Dictionaries.AsReadOnly(fields);

	public UnknownField this[int number]
	{
		get
		{
			if (!fields.TryGetValue(number, out var value))
			{
				return UnknownField.DefaultInstance;
			}
			return value;
		}
	}

	public int SerializedSize
	{
		get
		{
			int num = 0;
			foreach (KeyValuePair<int, UnknownField> field in fields)
			{
				num += field.Value.GetSerializedSize(field.Key);
			}
			return num;
		}
	}

	public int SerializedSizeAsMessageSet
	{
		get
		{
			int num = 0;
			foreach (KeyValuePair<int, UnknownField> field in fields)
			{
				num += field.Value.GetSerializedSizeAsMessageSetExtension(field.Key);
			}
			return num;
		}
	}

	private UnknownFieldSet(IDictionary<int, UnknownField> fields)
	{
		this.fields = fields;
	}

	public static Builder CreateBuilder()
	{
		return new Builder();
	}

	public static Builder CreateBuilder(UnknownFieldSet original)
	{
		return new Builder().MergeFrom(original);
	}

	public bool HasField(int field)
	{
		return fields.ContainsKey(field);
	}

	public void WriteTo(CodedOutputStream output)
	{
		foreach (KeyValuePair<int, UnknownField> field in fields)
		{
			field.Value.WriteTo(field.Key, output);
		}
	}

	public override string ToString()
	{
		return TextFormat.PrintToString(this);
	}

	public ByteString ToByteString()
	{
		ByteString.CodedBuilder codedBuilder = new ByteString.CodedBuilder(SerializedSize);
		WriteTo(codedBuilder.CodedOutput);
		return codedBuilder.Build();
	}

	public byte[] ToByteArray()
	{
		byte[] array = new byte[SerializedSize];
		CodedOutputStream codedOutputStream = CodedOutputStream.CreateInstance(array);
		WriteTo(codedOutputStream);
		codedOutputStream.CheckNoSpaceLeft();
		return array;
	}

	public void WriteTo(Stream output)
	{
		CodedOutputStream codedOutputStream = CodedOutputStream.CreateInstance(output);
		WriteTo(codedOutputStream);
		codedOutputStream.Flush();
	}

	public void WriteAsMessageSetTo(CodedOutputStream output)
	{
		foreach (KeyValuePair<int, UnknownField> field in fields)
		{
			field.Value.WriteAsMessageSetExtensionTo(field.Key, output);
		}
	}

	public override bool Equals(object other)
	{
		if (object.ReferenceEquals(this, other))
		{
			return true;
		}
		if (other is UnknownFieldSet unknownFieldSet)
		{
			return Dictionaries.Equals(fields, unknownFieldSet.fields);
		}
		return false;
	}

	public override int GetHashCode()
	{
		return Dictionaries.GetHashCode(fields);
	}

	public static UnknownFieldSet ParseFrom(CodedInputStream input)
	{
		return CreateBuilder().MergeFrom(input).Build();
	}

	public static UnknownFieldSet ParseFrom(ByteString data)
	{
		return CreateBuilder().MergeFrom(data).Build();
	}

	public static UnknownFieldSet ParseFrom(byte[] data)
	{
		return CreateBuilder().MergeFrom(data).Build();
	}

	public static UnknownFieldSet ParseFrom(Stream input)
	{
		return CreateBuilder().MergeFrom(input).Build();
	}
}
